src/hooks/UseRepositoryPermissions.ts

```diff
@@ -0,0 +1,96 @@
+import {useQuery} from '@tanstack/react-query';
+import {useState} from 'react';
+import {SearchState} from 'src/components/toolbar/SearchTypes';
+import {
+  fetchTeamRepoPermissions,
+  fetchUserRepoPermissions,
+  RepoMember,
+} from 'src/resources/RepositoryResource';
+import ColumnNames from 'src/routes/RepositoryDetails/Settings/ColumnNames';
+
+enum MemberType {
+  user = 'user',
+  robot = 'robot',
+  team = 'team',
+}
+
+export function useRepositoryPermissions(org: string, repo: string) {
+  const [page, setPage] = useState(1);
+  const [perPage, setPerPage] = useState(10);
+  const [search, setSearch] = useState<SearchState>({
+    query: '',
+    field: ColumnNames.account,
+  });
+
+  const {
+    data: userRoles,
+    isError: errorLoadingUserRoles,
+    isLoading: loadingUserRoles,
+  } = useQuery(
+    ['userrepopermissions', org, repo],
+    () => fetchUserRepoPermissions(org, repo),
+    {
+      placeholderData: {},
+    },
+  );
+
+  const {
+    data: teamRoles,
+    isError: errorLoadingTeamRoles,
+    isLoading: loadingTeamRoles,
+  } = useQuery(
+    ['teamrepopermissions', org, repo],
+    () => fetchTeamRepoPermissions(org, repo),
+    {
+      placeholderData: {},
+    },
+  );
+
+  const members: RepoMember[] = [];
+  for (const [name, roleData] of Object.entries(userRoles)) {
+    const type: MemberType = roleData.is_robot
+      ? MemberType.robot
+      : MemberType.user;
+    members.push({
+      org: org,
+      repo: repo,
+      name: name,
+      type: type,
+      role: roleData.role,
+    });
+  }
+  for (const [name, roleData] of Object.entries(teamRoles)) {
+    members.push({
+      org: org,
+      repo: repo,
+      name: name,
+      type: MemberType.team,
+      role: roleData.role,
+    });
+  }
+
+  const filteredMembers =
+    search.query !== ''
+      ? members?.filter((role) => role.name.includes(search.query))
+      : members;
+
+  const paginatedMembers = filteredMembers?.slice(
+    page * perPage - perPage,
+    page * perPage - perPage + perPage,
+  );
+
+  return {
+    loading: loadingUserRoles || loadingTeamRoles,
```

Comment:
You'll also want to check the value `isPlaceholderData` that is returned from the hook. When you use placeholder data, the the `loading` gets loaded as false and then switches to true once the request fires off.

Reply:
ahh interesting, updated.